### PR TITLE
Boundary traction

### PR DESCRIPTION
**Pull request by ilateur**
Add traction to the boundary conditions of the internal force calculation.

I'm not sure if it's finished yet. Maybe needs a test or an example? (It works for my rudimentary testing.)

### CODE REVIEW

File: src/physics/traction.hpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  VectorParameter posYside;
+  VectorParameter negYside;
+  VectorParameter posZside;
+  VectorParameter negZside;
+
+  /** Construct the DMI tensor for a given system */
+  explicit BoundaryTraction(std::shared_ptr<const System> system, std::string name = "");
+
+  /** Return CuDmiTensor */
+  CuBoundaryTraction cu() const;
```
Comment (DiegoDeGusem) [inline src/physics/traction.hpp:21]: I don't see how this is DMI?

**Review by DiegoDeGusem — COMMENTED** [review]
:tada:

File: src/physics/traction.hpp (hunk 1 above)
Comment (ilateur) [reply src/physics/traction.hpp:21]: left-over from copying DmiTensor as a template :face_with_peeking_eye:

File: src/physics/traction.hpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    if (sense == 1) { return posXside; }
+    else { return negXside; }
+  } else if (orientation == 1) {  // y
+    if (sense == 1) { return posYside; }
+    else { return negYside; }
+  } else {  // z
+    if (sense == 1) { return posZside; }
+    else { return negZside; }
+  }
+}
```
Comment (lamoreel) [inline src/physics/traction.hpp:53]: I _**think**_ this can be made a lot more efficient by using `orientation` and `sense` to calculate an index and use pointer arithmetic. With the members in the same order as you listed them above, this would become something like
```c++
  __device__ const CuVectorParameter& getSide(int orientation, int sense) const {
    const CuVectorParameter* base = &posXside;
    return *(base + (orientation * 2 + (1 - sense)));
  }
};
```
with the caveat that sense must be 0 (negative) or +1 (positive).

**Review by lamoreel — CHANGES_REQUESTED** [review]

**Comment by lamoreel** [comment]
Nice addition! And yes, there should be a test.

File: src/physics/traction.hpp (hunk 2 above)
Comment (lamoreel) [reply src/physics/traction.hpp:53]: I don't think you explicitly need an array. Since this is a `struct`, all members are laid out "next to each other" in memory (I think). This means you can do something like "`ptr_1 + 1 = ptr_2`".

File: src/physics/internalbodyforce.cu [hunk 3]
```diff
@@ -68,21 +70,24 @@ __global__ void k_internalBodyForce(CuField fField,
...
       );
     } else if (!ip1_inGeo) {
       // -11-- right boundary, custom difference + traction free BC,  ε ~ h^2
       f += ws[i] * (
         - 1./3. * stressTensor.vectorAt(coo_im1, stressRow)
         - stressTensor.vectorAt(idx, stressRow)  // -3/3 weight
-        // + 4./3. * real3{0, 0, 0}  // stress at coo_i+1/2 = boundary
+        // stress row at coo_i+1/2 = boundary traction * positive sense of normal vector
+        + 4./3. * traction.getSide(i, 1).vectorAt(idx)
       );
```
[the remark this answers is not in the document]
Comment (ilateur) [reply src/physics/internalbodyforce.cu:91]: - the stencil is -1 * stress at boundary, the stress is traction * normal vector component, which is -1 on the left/bottom/front. calculating -1*-1 is inefficient. The "negative sense of normal vector" comment was meant to make this "clear"
- oops

File: mumaxplus/magnet.py [hunk 4]
```diff
@@ -471,6 +472,16 @@ def rigid_shear_strain(self, value):
             raise Exception("Can not use shear strain with elastodynamics enabled.")
         self.rigid_shear_strain.set(value)
 
+    @property
+    def boundary_traction(self) -> BoundaryTraction:
+        """Get the boundary traction of this Magnet.
```
[the remark this answers is not in the document]
Comment (ilateur) [reply mumaxplus/magnet.py:477]: The dmi tensor property also has no unit :yum:

File: mumaxplus/traction.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
+"""BoundaryTraction implementation."""
+
+from .parameter import Parameter
+
+class BoundaryTraction:
+    """Contains the traction parameters of a Magnet at each of the 6 sides of
+    rectangular cells. The traction is applied as a boundary condition of the
+    stress tensor σ during calculation of the internal force.
+    
+    t = σ·n = n·σ
```
[the remark this answers is not in the document]
Comment (ilateur) [reply mumaxplus/traction.py:10]: It serves as a reminder that the stress tensor is symmetric, so the BC can be applied to either rows or columns, but no not really

File: mumaxplus/traction.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
+"""BoundaryTraction implementation."""
+
+from .parameter import Parameter
+
+class BoundaryTraction:
+    """Contains the traction parameters of a Magnet at each of the 6 sides of
+    rectangular cells. The traction is applied as a boundary condition of the
+    stress tensor σ during calculation of the internal force.
```
[the remark this answers is not in the document]
Comment (lamoreel) [reply mumaxplus/traction.py:8]: I mentioned in another comment that the third mentioning of "boundary" should be specified here, so 3 in total seems sufficient :smile:. The confusion arises because you write very generally "at each of the 6 sides of rectangular cells", but this is only applicable to some sides of some cells.

File: mumaxplus/traction.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+        warning
+        -------
+            The end user should not create BoundaryTraction instances. Each Magnet
+            already has BoundaryTraction as an attribute which can be used to set
+            the traction parameters. See :func:`Magnet.boundary_traction`.
+
+        Parameters
+        ----------
+        impl: _mumaxpluscpp.boundary_traction
```
[the remark this answers is not in the document]
Comment (ilateur) [reply mumaxplus/traction.py:31]: It's not because functions aren't meant for the average end-user that they shouldn't be documented.
(also again looking at `dmitensor.py`)

File: src/physics/internalbodyforce.cu [hunk 8]
```diff
@@ -19,11 +20,12 @@ __device__ int3 tensorRowComps(int row) {
  * Numerical divergence of stress with central five-point stencil in bulk material.
  * Lower order accuracy (three-point stencil) central difference is used in bulk
  * 1 cell away from boundary.
- * At the boundary, traction-free boundary conditions are implemented, using a
```
[the remark this answers is not in the document]
Comment (lamoreel) [reply src/physics/internalbodyforce.cu:22]: Bwa..

File: test/test_boundary_traction.py [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+        geom -= Ellipsoid(nx*cx/2, ny*cy/2, nz*cz/2)  # make hole
+        geom.translate((nx - 1)/2 * cx, (ny - 1)/2 * cy, (nz - 1)/2 * cz)  # to center
+
+        self.world = World((cx, cy, cz))
+        self.magnet = Ferromagnet(self.world, Grid((nx, ny, nz)), geometry=geom)
+        self.magnet.enable_elastodynamics = True
+        self.magnet.C11, self.magnet.C12, self.magnet.C44 = C11, C12, C44
+
+        # random traction to set
+        self.traction = np.random.uniform(low=-max_traction, high=max_traction, size=(3, nz, ny, nx))
```
Comment (lamoreel) [inline test/test_boundary_traction.py:30]: My only concern is that this distribution includes zero, although the possibility to get a (numerical) zero is vanishingly small.

**Review by lamoreel — APPROVED** [review]

File: test/test_boundary_traction.py (hunk 9 above)
Comment (ilateur) [reply test/test_boundary_traction.py:30]: I think `np.allclose` should be able to handle very small numbers :)

**Review by DiegoDeGusem — APPROVED** [review]
